Bukkit/0087-Add-PlayerSettingsUpdateEvent.patch

```diff
@@ -0,0 +1,389 @@
++                .append(chatColors)
++                .append(skinParts)
++                .append(mainHand)
++                .toHashCode();
++    }
++
++    /**
++     * Represents a delta-change from a previous player setting.
++     * - Any non-null fields will be replaced.
++     * - Any null fields will use the value from the old settings.
++     */
++    public static class Delta extends PlayerSettings {
++
++        public Delta(PlayerSettings old, String locale, Integer renderDistance, ChatVisibility chatVisibility, Boolean chatColors, Set<Skin.Part> skinParts, MainHand mainHand) {
++            super(
++                locale != null ? locale : old.getRawLocale(),
++                renderDistance != null ? renderDistance : old.getRenderDistance(),
++                chatVisibility != null ? chatVisibility : old.getChatVisibility(),
++                chatColors != null ? chatColors : old.getChatColors(),
++                skinParts != null ? skinParts : old.getSkinParts(),
++                mainHand != null ? mainHand : old.getMainHand()
++            );
++        }
++
++    }
++
++    /**
++     * Represents settings that have not been received from the player
++     * or have been manually cleared.
++     */
++    public static class Empty extends PlayerSettings {
++
++        public Empty() {
++            super(null, null, null, null, null, null);
++        }
++
++    }
++
++}
+diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
+index acad785..1786910 100644
+--- a/src/main/java/org/bukkit/entity/Player.java
++++ b/src/main/java/org/bukkit/entity/Player.java
+@@ -15,6 +15,7 @@ import org.bukkit.Material;
+ import org.bukkit.Note;
+ import org.bukkit.OfflinePlayer;
+ import org.bukkit.Particle;
++import org.bukkit.PlayerSettings;
+ import org.bukkit.Skin;
+ import org.bukkit.Sound;
+ import org.bukkit.Statistic;
+@@ -1643,4 +1644,28 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
+      * If the given item type is not cooling down, 0 is returned.
+      */
+     float getRemainingItemCooldown(Material item);
++
++    /**
++     * Get the various client-established settings of this player.
++     *
++     * If the player has not sent these values to the server,
++     * each getter, with the exclusion of {@link #getLocale()},
++     * will return null.
++     *
++     * @return the player's client settings.
++     */
++    PlayerSettings getSettings();
++
++    /**
++     * Forcibly set the player's client-established settings.
++     *
++     * This is recommended if the server has previous knowledge
++     * of a player's settings (like being stored in a database).
++     *
++     * Unlike when the player sends the packet to the server,
++     * this method does not call {@link org.bukkit.event.player.PlayerSettingsUpdateEvent}.
++     *
++     * @param settings the new player settings.
++     */
++    void setSettings(PlayerSettings settings);
```

Comment:
Would you ever want to call this method after receiving the client settings packet? That would just desync the server and client. Maybe this method should do nothing if the packet has already been received?